### PR TITLE
Utils finished

### PR DESCRIPTION
**Pull request by AlanCui4080**
To prevent building from failure caused by more restricted checks brought by newer GCC 12 that not contained in GCC 8, i removed it from Makefile.
e.g. -Waddress -Wmayve-uninitialized -Winfinite-recursion

All commands passed tests.

Alan.

### CODE REVIEW

File: user/find.c [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    if (this_sdata.type == T_DIR)
+      find(this_name, filename);
+  next:
+    close(this_fd);
+    ssize += sizeof(this_dirdata);
+  }
+fail:
+  close(dir_fd);
+  return;
+}
```
Comment (inclyc) [inline user/find.c:86]: \n

Have you completely forgotten what I said about adding a newline at the end of the text?

File: user/sleep.c [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+#include "user/user.h"
+
+int main(int argc, char *argv[]) {
+  if (argc != 2) {
+    fprintf(2, "The number of arugment(s) passed is invaild \n"
+               "Usage: sleep [ticks] \n");
+    exit(-1);
+  }
+  exit(sleep(atoi(argv[1])));
+}
```
Comment (inclyc) [inline user/sleep.c:11]: \n

File: user/xargs.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    int sub_rtn;
+    wait(&sub_rtn);
+    break;
+  default:
+    int rtn_exec = exec(argv[1], exec_argv);
+    fprintf(stderr, "Failed to exec() %d\n", rtn_exec);
+    exit(-1);
+  }
+  exit(0);
+}
```
Comment (inclyc) [inline user/xargs.c:59]: \n

File: user/find.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  uint64 ssize = 0;
+  while (ssize < sdata.size) {
+    struct dirent this_dirdata;
+    rtn = read(dir_fd, &this_dirdata, sizeof(struct dirent));
+    if (rtn < 0) {
+      fprintf(2, "Failed to read %s: %d\n", dirname, rtn);
+      goto fail;
+    }
+    if (!strlen(this_dirdata.name))
+      goto next;
```
Comment (inclyc) [inline user/find.c:47]: Don't use this f*king goto. Are you torturing your reviewer? What's this spaghetti code. Your code is not only unreadable for humans, it is even a disaster for C compilers! Can't you try to split the control flow into other functions? Why do you like troublesome goto so much? Your code has a serious misuse of goto!

File: user/pingpong.c [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  
+  int fork_rtn = fork();
+  switch (fork_rtn) {
+  case -1:
+    fprintf(2, "Failed to fork(): %d\n", fork_rtn);
+  fail:
+    close(ping_fd[1]);
+    close(pong_fd[0]);
+    close(pong_fd[1]);
+    break;
```
Comment (inclyc) [inline user/pingpong.c:31]: fail, but exit(0) ?

File: user/xargs.c [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  }
+  for (; i < remain_argc - 1; i++) {
+    char * gets_rtn = gets(exec_argv[i], MAX_ARG_LEN-1);
+    unsigned long len = strlen(gets_rtn);
+    gets_rtn[len - 1] = '\0';
+    if(!len)
+      break;
+  }
+  exec_argv[i] = (void*) 0;
+  // for(unsigned long q = 0; q < MAXARG; q++){
```
Comment (inclyc) [inline user/xargs.c:41]: what is this? debugging comments?

File: user/find.c [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    if (this_sdata.type == T_DIR)
+      find(this_name, filename);
+  next:
+    close(this_fd);
+    ssize += sizeof(this_dirdata);
+  }
+fail:
+  close(dir_fd);
```
Comment (inclyc) [inline user/find.c:84]:
```
user/find.c: In function 'find':
user/find.c:84:3: warning: 'dir_fd' may be used uninitialized in this function [-Wmaybe-uninitialized]
   84 |   close(dir_fd);
      |   ^~~~~~~~~~~~~
user/find.c:80:5: warning: 'this_fd' may be used uninitialized in this function [-Wmaybe-uninitialized]
   80 |     close(this_fd);
      |     ^~~~~~~~~~~~~~
```
Can I beg you to compile your code before submitting it, and apply for a review without a warning?

File: user/xargs.c [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    dst[i] = src[i];
+  return 0;
+}
+int main(int argc, char *argv[]) {
+  static char __buf_exec_argv[MAXARG * MAX_ARG_LEN];
+  static char *exec_argv[MAXARG];
+  for (unsigned long i = 0; i < MAXARG; i++){
+    exec_argv[i] = __buf_exec_argv + i*MAX_ARG_LEN;
+  }
+    // exec_argv[i] = malloc(128);
```
Comment (inclyc) [inline user/xargs.c:20]: Don't comment codes.

File: user/xargs.c [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+#define stderr 2
+
+int static lstrcpy(char *dst, const char *src) {
+  for (unsigned long i = 0; i < strlen(src) + 1; i++)
+    dst[i] = src[i];
+  return 0;
+}
+int main(int argc, char *argv[]) {
+  static char __buf_exec_argv[MAXARG * MAX_ARG_LEN];
```
Comment (inclyc) [inline user/xargs.c:15]: **The use of two underscores (` __ ') in identifiers is reserved for the compiler's internal use according to the ANSI-C standard**. That's what you said *compatibility*?

File: user/primes.c [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  };
+
+  if (current_next)
+    do_output(n2);
+  wait(0);
+
+fail:
+  close(read_fd);
+  close(write_fd);
+  exit(0);
```
Comment (inclyc) [inline user/primes.c:98]:
```suggestion
  exit(1);
```

File: user/primes.c [hunk 11]
```diff
@@ -0,0 +1,99 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+#define PIPE_READ (0)
+#define PIPE_WRITE (1)
+
+#define TRUE (1)
```
Comment (inclyc) [inline user/primes.c:7]: Why you need parentheses?

File: user/primes.c [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+#define FALSE (0)
+
+#define TERMINATION (0xf00f00ff)
```
Comment (inclyc) [inline user/primes.c:10]: Why you need parentheses?

File: user/primes.c [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    }
+  }
+  return;
+}
+void do_sub(void) {
+  int read_fd = 0;
+  int write_fd = 0;
+  int n1, n2;
+
+  read_fd = current[PIPE_READ];
```
Comment (inclyc) [inline user/primes.c:63]:
```suggestion
  int read_fd = current[PIPE_READ];
```



Could you please just declare `read_fd` here? Why can't you get f**king variables declared as they are used?

File: user/primes.c [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+void do_output(int num) {
+  write(current[PIPE_WRITE], &num, sizeof(num));
+
+  if (!current_next) {
+    int fork_rtn = fork();
+    switch (fork_rtn) {
+    case -1:
+      fprintf(2, "Failed to fork: %d\n", fork_rtn);
+      exit(-1);
+      break;
```
Comment (inclyc) [inline user/primes.c:44]: break after exit(?

This is totally unnecessary.

File: user/xargs.c [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+  if (argc < 2 || argc > MAXARG) {
+    fprintf(2, "The number of arugment(s) passed is invaild \n"
+               "Usage: xargs command \n");
+    exit(-1);
+  }
+
+  int remain_argc = MAXARG - 1;
+  unsigned long i = 0;
+  for (i = 0; i < (argc - 1); i++) {
```
Comment (inclyc) [inline user/xargs.c:30]:
```suggestion
  for (i = 0; i < argc - 1; i++) {
```

File: user/xargs.c [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    lstrcpy(exec_argv[i], argv[i + 1]);
+  }
+  for (; i < remain_argc - 1; i++) {
+    char * gets_rtn = gets(exec_argv[i], MAX_ARG_LEN-1);
```
Comment (inclyc) [inline user/xargs.c:34]:
```suggestion
    char * gets_rtn = gets(exec_argv[i], MAX_ARG_LEN - 1);
```

Did you **formatted** your source file?

File: user/find.c [hunk 17]
```diff
@@ -0,0 +1,86 @@
+#include "kernel/types.h"
```
Comment (inclyc) [inline user/find.c:1]: This program is completely non-compliant and needs to be rewritten.

**Comment by inclyc** [comment]
What the fxxking you close your PR but requested review from me? Are you f**king kidding me???

**Comment by AlanCui4080** [comment]
Sorry, I found that my wrong operations on git actually mess almost everything up so made it closed.
And i used to treat closing PR as a simple way to clean all these up.
There will be no more one time like this, I promise.

**Comment by inclyc** [comment]
> Sorry, I found that my wrong operations on git actually mess almost everything up so made it closed. And i used to treat closing PR as a simple way to clean all these up. There will be no more one time like this, I promise.

You still need to read review feedback comments carefully. Don't make mistakes twice!

**Pull request closed without merging**